Wikipedia/Code/AppDelegate.m

```diff
@@ -52,11 +52,41 @@ - (BOOL)application:(UIApplication*)application didFinishLaunchingWithOptions:(N
     [vc launchAppInWindow:self.window];
     self.appViewController = vc;
 
+    [self updateDynamicIconShortcutItems];
+
     NSLog(@"%@", [[[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask] lastObject]);
 
     return YES;
 }
 
+- (void)updateDynamicIconShortcutItems {
+    self.appViewController.shortcutItemSelectedAtLaunch = nil;
+
+    if (![[UIApplication sharedApplication] respondsToSelector:@selector(shortcutItems)]) {
+        return;
+    }
+
+    NSMutableArray* shortcutItems =
+        [[NSMutableArray alloc] initWithObjects:
+         [UIApplicationShortcutItem wmf_shortcutItemOfType:WMFIconShortcutTypeRandom],
+         [UIApplicationShortcutItem wmf_shortcutItemOfType:WMFIconShortcutTypeNearby],
+         nil
+        ];
+
+    MWKTitle* lastRead = [[NSUserDefaults standardUserDefaults] wmf_openArticleTitle];
+    if (lastRead) {
+        [shortcutItems addObject:[UIApplicationShortcutItem wmf_shortcutItemOfType:WMFIconShortcutTypeContinueReading]];
+    }
+
+    [shortcutItems addObject:[UIApplicationShortcutItem wmf_shortcutItemOfType:WMFIconShortcutTypeSearch]];
+
+    [UIApplication sharedApplication].shortcutItems = shortcutItems;
+}
+
+- (void)application:(UIApplication*)application performActionForShortcutItem:(UIApplicationShortcutItem*)shortcutItem completionHandler:(void (^)(BOOL))completionHandler {
+    self.appViewController.shortcutItemSelectedAtLaunch = shortcutItem;
```

Comment:
so, while things seem to work, we're ignoring the `completionHandler` parameter, which is supposed to say whether or not the action was successful . 

i have no idea what the consequences of never calling this could be, so I'd rather be on the safe side and honor the API contract by calling the block.  also, the docs state:

> If you find that your app was indeed launched using a quick action, perform the requested quick action within the launch method and return a value of NO from that method. When you return a value of NO, the system does not call the application:performActionForShortcutItem:completionHandler: method.

However, our `performActionForShortcutItem:` isn't being called twice, even though we're returning `YES` from `didFinishLaunching:`.  Not going to block PR since it's functional, but I'd like to talk about this more tomorrow, since I'm now a bit confused :confounded: